avocado/core/nrunner.py

```diff
@@ -643,14 +644,28 @@ class Task:
     While a runnable describes what to be run, and gets run by a
     runner, a task should be a unique entity to track its state,
     that is, whether it is pending, is running or has finished.
-
-    :param identifier:
-    :param runnable:
     """
 
     def __init__(self, identifier, runnable, status_uris=None,
                  known_runners=None):
-        self.identifier = identifier
+        """Instantiates a new Task.
+
+        :param identifier: any identifier that is guaranteed to be unique
+                           within the context of a Job. A recommended value
+                           is a :class:`avocado.core.test_id.TestID` instance
+                           when a task represents a test, because besides the
+                           uniqueness aspect, it's also descriptive.  If an
+                           identifier is not given, an automatically generated
+                           one will be set.
+        :param runnable: the "description" of what the task should run.
+        :type runnable: :class:`avocado.core.nrunner.Runnable`
+        :param status_uri: the URIs for the status servers that this task
+                           should send updates to.
+        :type status_uri: list
+        :param known_runners: a mapping of runnable kinds to runners.
+        :type known_runners: dict
+        """
+        self.identifier = identifier or str(uuid1())
```

Comment:
I see your point about moving the definition of the `identifier` when the class is instantiated, removing it from the `utils.py` file, making it explicit, and I agree we should do it here.

I also agree with @richtja and @beraldoleal, and setting the parameter to `None` seems more reasonable to me.

Another point that came to my attention is that the task `identifier` is still changed after the object is created here: https://github.com/avocado-framework/avocado/blob/1dc84bd46950eba6c5d2bbd57d21528a7095014f/avocado/plugins/runner_nrunner.py#L195, as we discussed here https://github.com/avocado-framework/avocado/pull/4423#discussion_r594507205.

Changing the `identifier` after the object instantiation cancels any good intent to set it here. In this case, I keep my suggestion to decouple the task id from the test id.

Reply:
> I see your point about moving the definition of the `identifier` when the class is instantiated, removing it from the `utils.py` file, making it explicit, and I agree we should do it here.
> 
> I also agree with @richtja and @beraldoleal, and setting the parameter to `None` seems more reasonable to me.
> 

OK!

> Another point that came to my attention is that the task `identifier` is still changed after the object is created here:
> 
> https://github.com/avocado-framework/avocado/blob/1dc84bd46950eba6c5d2bbd57d21528a7095014f/avocado/plugins/runner_nrunner.py#L195
> 
> , as we discussed here [#4423 (comment)](https://github.com/avocado-framework/avocado/pull/4423#discussion_r594507205).
> Changing the `identifier` after the object instantiation cancels any good intent to set it here. In this case, I keep my suggestion to decouple the task id from the test id.

I don't think so. We're not mandating that the Task identifier should never be changed (and that's made clearer in the docstring added here. In theory, using the Job API, I can create my own test suite, with my own Tasks.  If I have two jobs on a single Python scripts, I can have two Tasks with the same identifier (it may even make sense, for instance, to run a Job with a single CPU, and then another Job with multiple CPUs, so that we compare the test behaviors based on their identification).

So the uniqueness requirement is one that the state machine / status server have.  It may be a good idea indeed to check that the uniqueness value they're using applies throughout the job -- as a sanity check.  With the recommendation to use the Python objects themselves in the *state machine* that's safe, but for the status server, it's not.

Other than that, I don't think we need another attribute on tasks at this point, nor should we try to define that policy on the Task itself (because it's not aware of the other Tasks in a Job).

Reply:
> > I see your point about moving the definition of the `identifier` when the class is instantiated, removing it from the `utils.py` file, making it explicit, and I agree we should do it here.
> > I also agree with @richtja and @beraldoleal, and setting the parameter to `None` seems more reasonable to me.
> 
> OK!
> 
> > Another point that came to my attention is that the task `identifier` is still changed after the object is created here:
> > https://github.com/avocado-framework/avocado/blob/1dc84bd46950eba6c5d2bbd57d21528a7095014f/avocado/plugins/runner_nrunner.py#L195
> > 
> > , as we discussed here [#4423 (comment)](https://github.com/avocado-framework/avocado/pull/4423#discussion_r594507205).
> > Changing the `identifier` after the object instantiation cancels any good intent to set it here. In this case, I keep my suggestion to decouple the task id from the test id.
> 
> I don't think so. We're not mandating that the Task identifier should never be changed (and that's made clearer in the docstring added here. In theory, using the Job API, I can create my own test suite, with my own Tasks. If I have two jobs on a single Python scripts, I can have two Tasks with the same identifier (it may even make sense, for instance, to run a Job with a single CPU, and then another Job with multiple CPUs, so that we compare the test behaviors based on their identification).
> 
> So the uniqueness requirement is one that the state machine / status server have. It may be a good idea indeed to check that the uniqueness value they're using applies throughout the job -- as a sanity check. With the recommendation to use the Python objects themselves in the _state machine_ that's safe, but for the status server, it's not.
> 
> Other than that, I don't think we need another attribute on tasks at this point, nor should we try to define that policy on the Task itself (because it's not aware of the other Tasks in a Job).

Ah, absolutely not. I think `decouple` was not the correct word here. I don't think another attribute will solve the problem. My idea was more related to keep consistency, so if the task is a test, the task id represents the test id, otherwise, the task id should not be changed.